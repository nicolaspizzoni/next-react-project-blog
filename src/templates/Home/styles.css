.container {
  max-width: 1200px;
  margin: 0 auto;
  padding: 30px;
  /* altura da pagina no navegador */
  min-height: 100vh;
  background: #eee;
}

.posts {
  display: grid;
  /* repeat para todos os elementos, auto-fill para preencher todo espa√ßo, minmax(min: 280px, max 1fr) */
  grid-template-columns: repeat(auto-fill, minmax(280px, 1fr));
  gap: 30px;
}

.button-container {
  min-height: 150px;
  display: flex;
  align-items: flex-end;
}

.input-content {
  margin-bottom: 30px;
  font-size: 30px;
  width: 100%;
  padding: 5px 8px;
}