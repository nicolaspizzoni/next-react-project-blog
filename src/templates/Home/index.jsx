import { useState, useEffect } from 'react';
import './styles.css';
import PostCard from '../../components/PostCard';
import { loadPosts } from '../../utils/load-posts';
import { Button } from '../../components/Button';

function Home() {
  const [posts, setPosts] = useState([]);
  const [allPosts, setAllPosts] = useState([]);
  const [page, setPage] = useState(0);
  const [postPerPage] = useState(18);
  const [searchValue, setSearchValue] = useState('');

  const noMorePosts = page + postPerPage >= allPosts.length;

  const filteredPosts = searchValue
    ? allPosts.filter((post) => post.title.toLowerCase().includes(searchValue.toLowerCase()))
    : posts;

  useEffect(() => {
    async function load() {
      await handleLoadPosts();
    }
    load();
  });

  // CALLBACK THIS.SETSTATE

  // handleClickTeste = () => {

  //   this.setState demora para executar pois o React faz a comparação da nova Virtual DOM com a antiga
  //   porém o código continua executando, então se o console estivesse fora da callback do setState
  //   ele mostraria o valor anterior a nova atualização do valor

  //   this.setState({ counter: counter + 1 }, () => {
  //     console.log(this.state.counter)
  //   })
  // }

  const handleLoadPosts = async () => {
    const postsAndPhotos = await loadPosts();
    setPosts(postsAndPhotos.slice(page, postPerPage));
    setAllPosts(postsAndPhotos);
  };

  const loadMorePosts = () => {
    const nextPage = page + postPerPage;

    // pegando próximos posts no intervalo do slice de allPosts
    const nextPosts = allPosts.slice(nextPage, nextPage + postPerPage);

    // jogando posts pegos no intervalo de allPosts dentro dos posts iniciais
    posts.push(...nextPosts);

    setPosts(posts);
    setPage(nextPage);
  };

  const handleSearch = (e) => {
    const { value } = e.target;
    setSearchValue(value);
  };

  return (
    <section className="container">
      <input
        type="search"
        onChange={handleSearch}
        value={searchValue}
        className="input-content"
        placeholder="Pesquisar"
      />
      <div className="posts">
        {filteredPosts.length > 0 ? (
          filteredPosts.map((post) => <PostCard post={post} key={post.id} />)
        ) : (
          <p>Não há posts para essa pesquisa</p>
        )}
      </div>
      <div className="button-container">
        {!searchValue && <Button text={'Load more'} onClick={loadMorePosts} disabled={noMorePosts} />}
      </div>
    </section>
  );
}

// class Home2 extends Component {

//   state = {
//     posts: [],
//     allPosts: [],
//     page: 0,
//     postPerPage: 18,
//     searchValue: '',
//   }

//   async componentDidMount(){
//     await this.loadPosts()
//   }

//   loadPosts = async () => {
//     const {page, postPerPage} = this.state

//     const postsAndPhotos = await loadPosts()
//     this.setState({
//       posts: postsAndPhotos.slice(page, postPerPage),
//       allPosts: postsAndPhotos,
//     })
//   }

//   loadMorePosts = () => {
//     const {
//       page,
//       postPerPage,
//       allPosts,
//       posts
//     } = this.state;
//     const nextPage = page + postPerPage;
//     const nextPosts = allPosts.slice(nextPage, nextPage + postPerPage)

//     posts.push(...nextPosts);

//     this.setState({
//       posts,
//       page: nextPage,
//     })
//   }

//   handleSearch = (e) => {
//     const {value} = e.target;
//     this.setState({searchValue: value})
//   }

//   render(){

//     const {posts, page, postPerPage, allPosts, searchValue} = this.state
//     const noMorePosts = page + postPerPage >= allPosts.length;

//     const filteredPosts = !!searchValue
//     ? allPosts.filter(post => post.title.toLowerCase().includes(searchValue.toLowerCase()))
//     : posts

//     return (
//       <section
//         className="container"
//       >
//         <input
//           type="search"
//           onChange={this.handleSearch}
//           value={searchValue}
//           className="input-content"
//           placeholder="Pesquisar"
//         />
//         <div className="posts">
//           {
//             filteredPosts.length > 0 ?
//               filteredPosts.map(post => (
//                 <PostCard post={post} key={post.id}/>
//               ))
//             : <p>Não há posts para essa pesquisa</p>
//           }
//         </div>
//         <div
//           className="button-container"
//         >
//           {
//             !searchValue &&
//             <Button
//               text={"Load more"}
//               onClick={this.loadMorePosts}
//               disabled={noMorePosts}
//             />
//           }
//         </div>
//       </section>
//     );
//   }
// }

export default Home;
