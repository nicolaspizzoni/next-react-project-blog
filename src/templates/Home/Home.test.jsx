import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Home from '.';

//describe define o conjunto de testes a serem feitos no componente
describe('<Home />', () => {
  test('dummy test 1', () => {
    expect(1).toBe(1);
  });

  test('dummy test 2', () => {
    expect(2).toBe(2);
  });

  // Falhou
  test('dummy test 3', () => {
    expect(3).toBe(3);
  });

  test('input', () => {
    render(<Home />);
    const input = screen.getByPlaceholderText(/Pesquisar/i);

    expect(input).toBeInTheDocument();

    expect(input.value).toBe('');

    const valor = 'Testando';

    // Criando evento de usuÃ¡rio de teclar no input o valor
    userEvent.type(input, valor);

    expect(input.value).toBe(valor);
    // expect(input.onchange).toHaveBeenCalledTimes(valor.length)
  });

  test('should match snapshot', () => {
    const { container } = render(<Home />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
