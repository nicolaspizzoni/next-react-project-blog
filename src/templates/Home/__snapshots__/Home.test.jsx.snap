// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should match snapshot 1`] = `
<section
  class="container"
>
  <input
    class="input-content"
    placeholder="Pesquisar"
    type="search"
    value=""
  />
  <div
    class="posts"
  >
    <p>
      Não há posts para essa pesquisa
    </p>
  </div>
  <div
    class="button-container"
  >
    <button
      class="button"
      disabled=""
    >
      Load more
    </button>
  </div>
</section>
`;
